### PR TITLE
Refactor `sc_executor::RuntimeInfo` trait into 2 parts

### LINKED ISSUE #9475: Small sc-executor refactorings
A couple of things that I've noticed recently that should be changed:

* `CodeExecutor` is not implemented for `WasmExecutor`, only `NativeExecutor`. This should be changed.
* `RuntimeInfo` should be split into 2 traits, `NativeVersion` and `WasmVersion` so that it can be implemented on both `NativeExecutor` and `WasmExecutor` sensibly.
* `new_full_client`/`new_full_client` in `sc-service` shouldn't force the use of a native executor with `NativeExecutionDispatch`.

I'm happy to take on this issue, but I'm busy for the time being.

### PR DESCRIPTION
**Pull request by expenses**
Partially closes #9475.

### CODE REVIEW

File: client/executor/src/lib.rs [hunk 1]
```diff
@@ -64,6 +61,12 @@ pub trait RuntimeInfo {
 	) -> error::Result<RuntimeVersion>;
 }
 
+/// Provides native runtime information.
+pub trait NativeRuntimeInfo {
```
Comment (bkchr) [inline client/executor/src/lib.rs:65]: This is the same trait as in `sp-version`, just a different name.

File: client/executor/src/native_executor.rs [hunk 2]
```diff
@@ -269,6 +269,48 @@ impl sp_core::traits::ReadRuntimeVersion for WasmExecutor {
...
+	fn call<
+		R: Decode + Encode + PartialEq,
+		NC: FnOnce() -> result::Result<R, Box<dyn std::error::Error + Send + Sync>> + UnwindSafe,
+	>(
+		&self,
+		ext: &mut dyn Externalities,
+		runtime_code: &RuntimeCode,
+		method: &str,
+		data: &[u8],
+		_use_native: bool,
```
Comment (gilescope) [inline client/executor/src/native_executor.rs:284]: if use_native is true then panic?

**Comment by gilescope** [comment]
Seems a step in the right direction to being native free. Not sure why the polkadot build is failing - I don't see how it's related to these changes?

File: client/executor/src/native_executor.rs [hunk 3]
```diff
@@ -269,6 +269,52 @@ impl sp_core::traits::ReadRuntimeVersion for WasmExecutor {
...
+		ext: &mut dyn Externalities,
+		runtime_code: &RuntimeCode,
+		method: &str,
+		data: &[u8],
+		use_native: bool,
+		_native_call: Option<NC>,
+	) -> (Result<NativeOrEncoded<R>>, bool) {
+		if use_native {
+			log::warn!("`use_native` is set to `true` when calling the `WasmExecutor`. Ignoring.");
+		}
```
Comment (bkchr) [inline client/executor/src/native_executor.rs:289]: And what should the user do with this?

Please remove this. It is not correct to print this here.

**Comment by expenses** [comment]
`continuous-integration/gitlab-unleash-check` seems to be failing for unrelated reasons.

**Review by pepyakin — APPROVED** [review]
Looks reasonable in isolation. I am looking forward to see how this will play out in the later phases of refactor.

**Review by gilescope — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Comment by expenses** [comment]
bot merge

**Pull request merged**